### PR TITLE
The Great Swift 4 Refactor

### PR DESCRIPTION
**Pull request by AttilaTheFun**
This PR addresses https://github.com/AttilaTheFun/SwaggerParser/issues/3 and refactors the library to remove the ObjectMapper dependency in favor of the native Codable protocol. It also includes some other smaller improvements like a nicer Version API and it leverages some of the dictionary methods built in to the standard library in Swift 4.

### CODE REVIEW

File: Sources/Enums.swift [hunk 1]
```diff
@@ -1,19 +1,26 @@
-import ObjectMapper
 
-public enum OAuth2FlowType: String {
+public enum ParameterLocation: String, Codable {
+    case query = "query"
+    case header = "header"
+    case path = "path"
+    case formData = "formData"
+    case body = "body"
```
Comment (n8chur) [inline Sources/Enums.swift]: Is it necessary to set these explicitly to their string values if the enum is `: String`? Why not omit the string value if the value's name matches the string representation?

File: Sources/Enums.swift [hunk 2]
```diff
@@ -79,24 +86,35 @@ public enum DataType: String {
...
-        if let typeString: String = try? map.value("type"), let mappedType = DataType(rawValue: typeString) {
-            self = mappedType
-        } else if map.JSON["$ref"] != nil {
+    public init(from decoder: Decoder) throws {
+        let values = try decoder.container(keyedBy: CodingKeys.self)
+        if values.contains(.type) {
+            guard let typeString = try? values.decode(String.self, forKey: .type),
+                let type = DataType(rawValue: typeString) else
+            {
+                throw DecodingError("Unknown data type")
```
Comment (n8chur) [inline Sources/Enums.swift]: Should the `typeString` be included in the `DecodingError`?

File: Sources/FixedParameterFields.swift [hunk 3]
```diff
@@ -29,24 +20,35 @@ public struct FixedParameterFields {
...
     }
 
+    init(from decoder: Decoder) throws {
+        let values = try decoder.container(keyedBy: CodingKeys.self)
+        self.name = try values.decode(String.self, forKey: .name)
+        self.location = try values.decode(ParameterLocation.self, forKey: .location)
+        self.description = try? values.decode(String.self, forKey: .description)
+        self.required = (try values.decodeIfPresent(Bool.self, forKey: .required)) ?? false
+        self.example = try? values.decode(String.self, forKey: .example)
+        // TODO: Decode example as any.
```
Comment (n8chur) [inline Sources/FixedParameterFields.swift:45]: What's preventing this from being done now?

File: Sources/Metadata.swift [hunk 4]
```diff
@@ -24,32 +23,58 @@ public struct Metadata {
...
-
-    typealias Building = Metadata
-
+struct MetadataBuilder: Codable {
     let type: DataType
     let title: String?
     let description: String?
     let defaultValue: Any?
-    let enumeratedValues: [Any?]?
+    let enumeratedValues: [Any]?
```
Comment (n8chur) [inline Sources/Metadata.swift]: Is nil not a valid value?

File: Sources/Metadata.swift [hunk 5]
```diff
@@ -24,32 +23,58 @@ public struct Metadata {
...
+
+    func encode(to encoder: Encoder) throws {
+        var container = encoder.container(keyedBy: CodingKeys.self)
+        try container.encode(self.title, forKey: .title)
+        try container.encode(self.description, forKey: .description)
+        // defaultValue
+        // enumeratedValues
+        try container.encode(self.nullable, forKey: .nullable)
+        // example
+        // TODO: Encode any values.
```
Comment (n8chur) [inline Sources/Metadata.swift]: Will this be completed in this PR?

File: Sources/ObjectSchema.swift [hunk 6]
```diff
@@ -10,64 +12,43 @@ public struct ObjectSchema {
...
-    public let minProperties: Int?
-
-    /// The maximum number of properties. If set it must be a non-negative integer.
-    public let maxProperties: Int?
-
     /// The additionalProperties keyword is used to control the handling of extra stuff, 
     /// that is, properties whose names are not listed in the properties keyword. 
     /// By default any additional properties are allowed.
-    /// The additionalProperties may be either a boolean or a schema.
+    /// The additionalProperties may bdde either a boolean or a schema.
```
Comment (n8chur) [inline Sources/ObjectSchema.swift]: Typo

File: Sources/Operation.swift [hunk 7]
```diff
@@ -44,52 +43,83 @@ public struct Operation {
...
+        try container.encode(self.deprecated, forKey: .deprecated)
+        try container.encode(self.identifier, forKey: .identifier)
+        try container.encode(self.tags, forKey: .tags)
+        try container.encode(self.security, forKey: .security)
+        try container.encode(self.externalDocumentationBuilder, forKey: .externalDocumentation)
+
+        try container.encode(self.parameters, forKey: .parameters)
+        var allResponses = [String: Reference<ResponseBuilder>]()
+        allResponses["default"] = self.defaultResponse
+        self.responses.forEach { allResponses[$0.description] = $1 }
```
Comment (n8chur) [inline Sources/Operation.swift]: Is `<Int>.description` preferable over `String(<Int>)`?

File: Sources/Version.swift [hunk 8]
```diff
@@ -1,39 +1,131 @@
...
 
         let components = string.components(separatedBy: ".")
-        guard components.count >= 1, let major = Int(components[0]) else {
-            return nil
+        let integerComponents = try components.map { component -> UInt in
+            if let int = UInt(component) {
+                return int
+            }
+
+            throw DecodingError("Unable to parse version component into integer")
```
Comment (n8chur) [inline Sources/Version.swift]: Should we add the `component` to this error string to assist in debugging?

File: Sources/Version.swift [hunk 9]
```diff
@@ -1,39 +1,131 @@
...
+        guard let lastComponent = integerComponents.last else {
+            throw DecodingError("Version components were empty")
+        }
+
+        let version = integerComponents.dropLast().reversed().reduce(Version.version(lastComponent))
+        { subversion, component in
+            return Version.subversion(component, subversion)
+        }
+
+        self = version
```
Comment (n8chur) [inline Sources/Version.swift]: Why not set self directly above?:
```
self = integerComponents.dropLast().reversed().reduce(Version.version(lastComponent))
...
```

File: Sources/Version.swift [hunk 10]
```diff
@@ -1,39 +1,131 @@
...
+    public static func ==(left: Version, right: Version) -> Bool {
+        let leftArray = left.map { $0.value }
+        let rightArray = right.map { $0.value }
+        return leftArray == rightArray
+    }
 }
+
+extension Version: Comparable {
+    public static func <(left: Version, right: Version) -> Bool {
+        let leftArray = left.map { $0.value }
```
Comment (n8chur) [inline Sources/Version.swift]: Should there be a `components` property on `Version`?

File: Sources/Version.swift [hunk 11]
```diff
@@ -1,39 +1,131 @@
...
+        case .version:
             return nil
         }
+    }
+
+    public var major: UInt { return self.value }
+    public var minor: UInt? { return self.subversion?.value }
+    public var patch: UInt? { return self.subversion?.subversion?.value }
+
+    public init(from decoder: Decoder) throws {
```
Comment (n8chur) [inline Sources/Version.swift:28]: Should we add a convenience initializer that takes a `String` (that this initializer calls)? I imagine that could be useful for version comparisons by the consumer.

**Review by n8chur — APPROVED** [review]
Looks great! I can't believe how well `Codable` works for this stuff...

Thanks for doing this!

File: Sources/Version.swift (hunk 10 above)
Comment (AttilaTheFun) [reply Sources/Version.swift]: Sure

File: Sources/Version.swift (hunk 11 above)
Comment (AttilaTheFun) [reply Sources/Version.swift:28]: Sure

File: Sources/FixedParameterFields.swift (hunk 3 above)
Comment (AttilaTheFun) [reply Sources/FixedParameterFields.swift:45]: I had to write extensions on Keyed(Encoding|Decoding)Container to allow encoding of Any types. Normally the encode/decode functions only take either a generic param conforming to the respective encoding/decoding protocol, or one of a limited set of predefined primitive types. (E.g. String, Int64 etc). I've done this now, though, if you want to take a look.

**Comment by AttilaTheFun** [comment]
@n8chur Okay, I think I've addressed all of your comments. @yonaskolb would you like to take a look? I'm planning on merging this and cutting a release after Xcode 9 GM goes live next week.

**Pull request merged**